tests/test_run_plan.py

```diff
@@ -0,0 +1,445 @@
+        docker=None,
+        build=None,
+        deployment=None,
+        dependencies=None,
+    )
+    if cached:
+        return ProjectExecution.skip(project)
+    return ProjectExecution.run(project)
+
+
+execution_1 = stub_execution("project 1")
+execution_2 = stub_execution("project 2")
+cached_execution = stub_execution("a cached project", cached=True)
+
+
+class TestEmptyPlan:
+    run_plan = RunPlan.empty()
+
+    def test_empty(self):
+        assert self.run_plan.all_known_projects == set()
+        assert not self.run_plan._full_plan
+        assert not self.run_plan.selected_plan
+
+    @pytest.mark.parametrize(
+        argnames="use_full_plan",
+        argvalues=[
+            pytest.param(True, id="use_full_plan=True"),
+            pytest.param(False, id="use_full_plan=False"),
+        ],
+    )
+    def test_get_all_projects(self, use_full_plan):
+        assert self.run_plan._get_all_projects(use_full_plan=use_full_plan) == set()
+
+    @pytest.mark.parametrize(
+        argnames="stage,use_full_plan",
+        argvalues=[
+            pytest.param(
+                build_stage, True, id=f"stage={build_stage.name}, use_full_plan=True"
+            ),
+            pytest.param(
+                build_stage, False, id=f"stage={build_stage.name}, use_full_plan=False"
+            ),
+            pytest.param(
+                test_stage, True, id=f"stage={test_stage.name}, use_full_plan=True"
+            ),
+            pytest.param(
+                test_stage, False, id=f"stage={test_stage.name}, use_full_plan=False"
+            ),
+            pytest.param(
+                deploy_stage, True, id=f"stage={deploy_stage.name}, use_full_plan=True"
+            ),
+            pytest.param(
+                deploy_stage,
+                False,
+                id=f"stage={deploy_stage.name}, use_full_plan=False",
+            ),
+        ],
+    )
+    def test_get_projects_for_stage(self, stage, use_full_plan):
+        assert (
+            self.run_plan.get_projects_for_stage(
+                stage=stage, use_full_plan=use_full_plan
+            )
+            == set()
+        )
+
+    @pytest.mark.parametrize(
+        argnames="use_full_plan",
+        argvalues=[
+            pytest.param(True, id="use_full_plan=True"),
+            pytest.param(False, id="use_full_plan=False"),
+        ],
+    )
+    def test_get_all_stages(self, use_full_plan):
+        assert self.run_plan._get_all_stages(use_full_plan=use_full_plan) == set()
+
+    @pytest.mark.parametrize(
+        argnames="include_cached_projects",
+        argvalues=[
+            pytest.param(True, id="include_cached_projects=True"),
+            pytest.param(False, id="include_cached_projects=False"),
+        ],
+    )
+    def test_has_projects_to_run(self, include_cached_projects):
+        assert not self.run_plan.has_projects_to_run(
+            include_cached_projects=include_cached_projects,
+        )
+
+
+class TestFullRunPlan:
```

Comment:
would it be worth considering putting each class in its own file?

Reply:
Just to check: I thought the convention was to merely prefix test files with `test_`:
* `project.py` → `test_project.py`
* `run_plan.py` → `test_run_plan.py`
* `validation.py` → `test_validation.py`

And so on.

Reply:
Ah fair point, I forgot about that!